### PR TITLE
[WIP] Ports signals and components from /tg/

### PR DESCRIPTION
**Pull request by mikomyazaki**
Ports components from /tg/

Currently basic system is implemented.

To Do:
- [ ] Convert some existing systems to components to act as examples.
- [ ] Write a git wiki page explaining how to use them.
- [x] Replace existing expansions with equivalent components & remove expansions
- [ ] Figure out how many signals I should put in right now.

### CODE REVIEW

**Comment by mikomyazaki** [comment]
!wip

File: code/datums/components/multitool/circuitboards/circuitboards.dm [hunk 1]
```diff
@@ -1,4 +1,4 @@
-/datum/expansion/multitool/circuitboards/CanUseTopic(var/mob/user)
+/datum/component/multitool/circuitboards/CanUseTopic(var/mob/user)
```
Comment (ninjanomnom) [inline code/datums/components/multitool/circuitboards/circuitboards.dm]: You really should avoid component subtypes and use either multiple components/elements or add more configuration to the init

File: code/datums/elements/_element.dm [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/datum/element/Destroy(force)
+	if(!force)
+		return QDEL_HINT_LETMELIVE
+	SSdcs.elements_by_type -= type
+	return ..()
+
+//DATUM PROCS
+
+/// Finds the singleton for the element type given and attaches it to src
+/datum/proc/AddElement(eletype, ...)
```
Comment (ninjanomnom) [inline code/datums/elements/_element.dm]: https://github.com/tgstation/tgstation/pull/49098

File: code/game/atoms_init.dm [hunk 3]
```diff
@@ -40,6 +40,8 @@
 	if (update_icon_on_init)
 		queue_icon_update()
 
+	ComponentInitialize()
```
Comment (ninjanomnom) [inline code/game/atoms_init.dm]: I would leave this out, it's obsolete and we prefer not to use it in tg

File: code/__defines/dcs/flags.dm [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+/// old component is deleted (default)
+#define COMPONENT_DUPE_HIGHLANDER		0
+/// duplicates allowed
+#define COMPONENT_DUPE_ALLOWED			1
+/// new component is deleted
+#define COMPONENT_DUPE_UNIQUE			2
+/// old component is given the initialization args of the new
+#define COMPONENT_DUPE_UNIQUE_PASSARGS	4
+/// each component of the same type is consulted as to whether the duplicate should be allowed
+#define COMPONENT_DUPE_SELECTIVE		5
```
Comment (skull132) [inline code/__defines/dcs/flags.dm]: Is this a fuck-up or meant to be `DUPLE_ALLOWED | DUPE_SELECTIVE`? The usual convention with bitflags is to define combinational flags as an OR-statement using the previous defines, for clarity.

Comment (ninjanomnom) [reply code/__defines/dcs/flags.dm]: It's a mistake that doesn't matter because these aren't used as bitflags

**Comment by NonQueueingMatt** [comment]
Outdated

**Pull request closed without merging**